### PR TITLE
Feature/event filters

### PR DESCRIPTION
**Pull request by laij84**

### CODE REVIEW

File: src/lib/debounce.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+module.exports = function debounce(fn, delay) {
```
Comment (ds300) [inline src/lib/debounce.js]: `yarn add lodash.debounce`?

File: src/components/events/eventListingCard.js [hunk 2]
```diff
@@ -11,24 +12,49 @@ const Card = styled(Link)`
...
+  opacity: 0;
 `
 
 const CardBody = styled.div`
   padding: 20px;
   background-color: ${props => props.theme.colors.white};
   flex-grow: 1;
 
-  @media (min-width: ${props => props.theme.breakpoints[0]}) {
+  ${media.mobile`
```
Comment (ds300) [inline src/components/events/eventListingCard.js]: this shouldn't be necessary. Mobile is the default. You can just put `padding: 30px` at the top

File: src/components/events/eventListingCard.js [hunk 3]
```diff
@@ -51,31 +77,49 @@ const CardPrice = styled.div`
...
     <Card to={`/events/${event.id}`}>
-      <CardImage
+      <CardImageWrapper
         src={`${event.eventsListPicture.file.url}?fit=fill&w=400&h=225&f=faces`}
-        alt={event.eventsListPicture.title}
-        width="400"
-        height="225"
-      />
+      >
+        <CardImage
```
Comment (ds300) [inline src/components/events/eventListingCard.js]: Seems like you're displaying the image twice, once as an `img` and once as the background on a `div`? Is there a reason for that?

File: src/components/events/eventListingCard.js [hunk 4]
```diff
@@ -51,31 +77,49 @@ const CardPrice = styled.div`
...
+          height="225"
+        />
+      </CardImageWrapper>
       <CardBody>
         <CardDate>{formatDate(event)}</CardDate>
         <CardHeading>{event.name}</CardHeading>
       </CardBody>
-      <CardPrice>from £{event.eventPriceLow}</CardPrice>
+      <CardPrice>
+        {event.isFree ? 'Free' : `from £${event.eventPriceLow}`}
```
Comment (ds300) [inline src/components/events/eventListingCard.js]: the `isFree` field has been deprecated. We should be using `eventPriceLow` and `eventPriceHigh` to determine whether an event is free or not.

File: src/theme/media.js [hunk 5]
```diff
@@ -2,6 +2,11 @@ import { css } from 'styled-components'
 import theme from './theme'
 
 export const media = {
+  mobile: (...args) => css`
```
Comment (ds300) [inline src/theme/media.js]: This should probably not exist. All styles apply to mobile unless specifically defined for higher screen widths.
```js
styled.blah`
   ${media.mobile`xyz`}
`
```
is functionally the same as
```js
styled.blah`
  xyz
`
```

File: src/theme/theme.js [hunk 6]
```diff
@@ -27,7 +27,7 @@ const theme = {
     body: 'Roboto, sans-serif',
   },
   space: [0, 5, 10, 15, 20],
-  breakpoints: ['375px', '768px', '1024px', '1440px'],
+  breakpoints: ['375px', '768px', '1280px', '1440px'],
```
Comment (ds300) [inline src/theme/theme.js]: Why was this changed? I think it makes more sense at 1024, and the web site looks better for screen sizes between 1024 and 1280 when it's at 1024.

File: src/pages/events.js [hunk 7]
```diff
@@ -17,69 +19,111 @@ const ColumnTextCenter = styled(Column)`
...
+class Events extends Component {
+  state = {
+    showFiltersMobile: false,
+  }
+
+  toggleFiltersMobile = () => {
+    const state = { ...this.state }
+    state.showFiltersMobile = !state.showFiltersMobile
+    this.setState(state)
+    noScroll.toggle()
```
Comment (ds300) [inline src/pages/events.js]:
```js
this.setState(state => ({
  ...state,
  showFiltersMobile: !state.showFiltersMobile,
}), () => noScroll.toggle())
```

File: src/pages/events.js [hunk 8]
```diff
@@ -17,69 +19,111 @@ const ColumnTextCenter = styled(Column)`
...
+            </Container>
+            <Container>
+              <Row>
+                <ColumnAddFilters width={1}>
+                  <Button onClick={this.toggleFiltersMobile} primary>
+                    <img
+                      src={filterIcon}
+                      width="22"
+                      height="18"
+                      alt="Filters Icon"
```
Comment (ds300) [inline src/pages/events.js]: not sure this is necessary. Probably best to put `role="presentation"`

File: src/components/appContext/index.js [hunk 9]
```diff
@@ -33,7 +35,48 @@ const initialState = {
...
+  constructor() {
+    super()
+    this.state = {
+      ...initialState,
+      breakpoint: this.getCurrentBreakpoint(),
+    }
+  }
+
+  componentDidMount() {
+    if (typeof window !== 'undefined') {
```
[the remark this answers is not in the document]
Comment (laij84) [reply src/components/appContext/index.js]: This is an issue in Gatsby. When you run the `gatsby build` command for production it will throw an error as there is no window server-side.

File: src/components/events/eventListingCard.js (hunk 2 above)
Comment (laij84) [reply src/components/events/eventListingCard.js]: yes the default is the xs mobile size 0-374px and mobile 375 up. Necessary to cater for devices under 375px like the iphone5 which is 320px

File: src/components/events/eventListingCard.js (hunk 3 above)
Comment (laij84) [reply src/components/events/eventListingCard.js]: Yes this is an old front-end trick. You use the background image as it gives you better control of the positioning and sizing, as older browsers won't support the css property object-fit.

You also include an IMG element for Accessbility (so screen readers can still interpret the img element and alt tags), but also for SEO as search engines look for img elements rather than background images.

File: src/theme/media.js (hunk 5 above)
Comment (laij84) [reply src/theme/media.js]: see above, to cater for xs and s mobile sizes

File: src/theme/theme.js (hunk 6 above)
Comment (laij84) [reply src/theme/theme.js]: the design file specifies these sizes. Also, it breaks the grid as there is a min-width property on the container for this bp at 1100px and 1024 is less than that so it ends up cut-off the screen.

File: src/components/events/eventListingCard.js (hunk 2 above)
Comment (ds300) [reply src/components/events/eventListingCard.js]: Ah, ok. Didn't realise we were using different styles for such small screens 👍

File: src/components/appContext/index.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (ds300) [reply src/components/appContext/index.js]: Odd. I was under the impression that componentDidMount doesn't get called server-side. Maybe gatsby is doing some extra stuff.

File: src/lib/debounce.js (hunk 1 above)
Comment (laij84) [reply src/lib/debounce.js]: ah I didn't know that you couldnt install the debounce function separately from the entire Lodash library

File: src/theme/theme.js (hunk 6 above)
Comment (laij84) [reply src/theme/theme.js]: The grid is already flexbox based so shouldn't be an issue. I think we can prob set it to 1024 but also modify the the min-width for the container to be 960 so it doesn't break.

File: src/components/events/eventListingCard.js (hunk 2 above)
Comment (laij84) [reply src/components/events/eventListingCard.js]: There aren't designs for screens under 375, but as a standard 320 is usually the smallest size you build for as the small iphone is that size and quite widely used. When the screen gets down this small it does make sense to reduce things like padding to make the text more readable.
